api/src/main/java/io/strimzi/api/kafka/model/template/ServiceTemplate.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2018, Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.api.kafka.model.template;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import io.strimzi.crdgenerator.annotations.Description;
+import io.sundr.builder.annotations.Buildable;
+
+import java.io.Serializable;
+
+/**
+ * Representation of a Pod template for Strimzi services.
+ */
+@Buildable(
+        editableEnabled = false,
+        generateBuilderPackage = false,
+        builderPackage = "io.fabric8.kubernetes.api.builder"
+)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({
+        "metadata"})
+public class ServiceTemplate implements Serializable {
```

Comment:
Rather than having a `*Template` class for each kind of resource, couldn't you use a single generic `ResourceTemplate<T extends HasMetadata>` and then in the ZookeeperTemplate, for example you could have:

```
    private ResourceTemplate<StatefulSet> statefulset;
    private ResourceTemplate<Pod> pod;
    private ResourceTemplate<Service> clientService;
    private ResourceTemplate<Service> nodesService;
```

I admit I'm not 100% certain this will work with the serialization via Jackson, but I'm pretty sure it will.

Strictly speaking, I'm not even sure you need the generics, since all these resource templates are the same, there's no type safety to be gained AFAICS.

Reply:
How would this work if we want to  add more into these templates later? My expectation is that this can go beyond metadata. For example it could contain things like:
* ImagePullPolicy
* ImagePullSecrets

And we can consider moving there things like resources, affinity in the long term etc.

Reply:
We could add specific `PodTemplate` class at the time we did that. There would be no inheritance between `ResourceTemplate` and `PodTemplate`. Since the type info is not serialized to YAML directly it would be a compatible change.

Reply:
I'm not sure its worth it when we already have the separate classes for Pods and Statefulsets. Also, why would you want no inheritance there? We can have the ResourceTemplate with the metadata stuff and simply extend it with PodTemplate or StatefulSet template with some additional stuff, or?

Reply:
The reason to avoid inheritance is because it makes the API less typesafe -- it would allow someone to pass a `PodTemplate` where a `ResourceTemplate` was expected. I suppose this is a reason to use generics, so `PodTemplate extends ResourceTemplate<Pod>`; that would restore the type safety assuming setters used correctly instantiated types.